test/types/peg.test-d.ts

```diff
@@ -0,0 +1,330 @@
+import * as peggy from "../..";
+import tsd from "tsd";
+import {expectType} from "tsd";
+import formatter from "tsd/dist/lib/formatter";
+import {readFileSync} from "fs";
+import {join} from "path";
+
+
+const src = readFileSync(
+  join(__dirname, "..", "..", "examples", "fizzbuzz.peggy"),
+  "utf8"
+);
+
+describe("peg.d.ts", () => {
+  it("executes a grammar", () => {
+    expectType<string>(src);
+    expect(src.length).toBeGreaterThan(0);
+
+    const parser = peggy.generate(src);
+    expectType<peggy.Parser>(parser);
+
+    let res = parser.parse("1\n");
+    expectType<any>(res);
+    expect(res).toStrictEqual([1]);
+
+    res = parser.parse("buzz\n11\nfizz\n", { start: 10 });
+    expect(res).toStrictEqual([ "buzz", 11, "fizz" ]);
+  });
+
+  it("takes a valid tracer", () => {
+    const parser = peggy.generate(src, { trace: true });
+    expectType<peggy.Parser>(parser);
+
+    const res = parser.parse(" /**/ 1\n", {
+      startRule: "top",
+      tracer: {
+        trace(event) {
+          expectType<peggy.ParserTracerEvent>(event);
+          expectType<"rule.enter" | "rule.match" | "rule.fail">(event.type);
+          expectType<string>(event.rule);
+          switch (event.type) {
+            case "rule.enter":
+              expectType<peggy.LocationRange>(event.location);
+              break;
+            case "rule.match":
+              expectType<peggy.LocationRange>(event.location);
+              expectType<any>(event.result);
+              break;
+          }
+        }
+      }
+    })
+  });
+
+  it("takes an output and grammarSource", () => {
+    const p1 = peggy.generate(src, { output: "parser", grammarSource: "src" });
+    expectType<peggy.Parser>(p1);
+
+    const p2 = peggy.generate(src, { output: "source" });
+    expectType<string>(p2);
+  });
+
+  it("creates an AST", () => {
+    const grammar = peggy.parser.parse(src);
+    expectType<peggy.ast.Grammar>(grammar);
+
+    const visited: { [key: string]: number; } = {};
+    function add(typ:string): void {
+      if (!visited[typ]) {
+        visited[typ] = 1;
+      } else {
+        visited[typ]++
+      }
+    }
+
+    const visit = peggy.compiler.visitor.build({
+      grammar(node) {
+        add(node.type);
+        expectType<peggy.ast.Grammar>(node);
+        expectType<"grammar">(node.type);
+        expectType<peggy.LocationRange>(node.location);
+        expectType<peggy.ast.TopLevelInitializer|undefined>(node.topLevelInitializer);
+        expectType<peggy.ast.Initializer|undefined>(node.initializer);
+        expectType<peggy.ast.Rule[]>(node.rules);
```

Comment:
It is necessary to check types of `node` fields? Hasn't the whole `node` test checked them out yet?

Reply:
What I'm trying to achieve is that we'll notice if we break backward-compatibility in the future.

Reply:
Makes sense